### ISSUE TITLE
Unnecessary function evaluations in optimize.minimize/SLSQP

### ISSUE THREAD

**Issue by ChristophNa** [scipy/scipy#10738]
I am currently running an optimization where one function evaluation costs approx. 3.5 minutes so I am interested in having as few function calls as possible. I am monitoring every function evaluation and it seems to me that minimize calls my function twice with exactly the same parameter values. Maybe the first time is to determine the function value and the second time for the numerical computation of the jacobian?
Is it possible to eliminate this unnecessary function call by passing the current function value to the numerical jacobian routine?

**Comment by andyfaff** [comment]
It might be doable. Just checking, are you talking about the Jacobian of your objective, or of the constraints?

**Comment by TeamPoro** [comment]
I am talking about the Jacobian of the objective, I currently don't have constraints.

**Comment by fuglede** [comment]
Apologies if this is just noise (as it certainly offers nothing to solve the issue, and chances are you're already well aware of this being an option), but from the sound of it, until the underlying issue is resolved, maybe you can make do of [`functools.lru_cache`](https://docs.python.org/3/library/functools.html#functools.lru_cache) to speed up consecutive calls to your slow function? That is, rather than pass `slow_func` to `minimize`, you could pass `functools.lru_cache(maxsize=16)(slow_func)`.

**Comment by andyfaff** [comment]
I don't think that `lru_cache` will work on numpy arrays. You may be better off looking at [`joblib.Memory`](https://joblib.readthedocs.io/en/latest/memory.html)

**Comment by mchirsa5** [comment]
Hi Andy, I would like to ask for how to find a cache directory for "function evaluation"? I try to use "dir()" in "minimize" but I can't find any "function evaluation" there.

**Comment by amarvin** [comment]
If someone's good with Fortran, I believe the fix must be done in https://github.com/scipy/scipy/blob/master/scipy/optimize/slsqp/slsqp_optmz.f

**Comment by amarvin** [comment]
Nevermind, the fix must be in Python and not Fortran. I believe that [`approx_jacobian`](https://github.com/scipy/scipy/blob/47ffc1ef2480df307392670a5476701b57e5333c/scipy/optimize/slsqp.py#L31) will need an optional argument for providing f0, the function value at x0. Somehow the f0 needs to passed to [the wrapping call](https://github.com/scipy/scipy/blob/47ffc1ef2480df307392670a5476701b57e5333c/scipy/optimize/slsqp.py#L305) as well.

**Comment by andyfaff** [comment]
This might be addressed by #10673

**Comment by andyfaff** [comment]
This seems to be fixed in master:

```
from scipy.optimize import minimize, rosen
import numpy as np

class F():
    def __init__(self):
        self.nfev = 0
        self.x_hist = []

    def __call__(self, x):
        self.nfev += 1
        self.x_hist.append(x)
        return rosen(x)

f = F()


minimize(f, (1.2, 1.2), method='SLSQP')

dup = 0
for i, e in enumerate(f.x_hist):
    same = np.array_equal(f.x_hist[i], f.x_hist[i - 1])
    if same:
        dup += 1
    print(e, same)
print(dup)
```

gives:

```
[1.2 1.2] False
[1.20000001 1.2       ] False
[1.2        1.20000001] False
[-114.4000089    49.19999851] False
[-10.36000089   5.99999985] False
[0.04399991 1.67999999] False
[0.99104274 1.28676425] False
[1.10694929 1.23863697] False
[1.1069493  1.23863697] False
[1.10694929 1.23863698] False
[1.00846235 0.99041619] False
[1.07203482 1.15064057] False
[1.07203484 1.15064057] False
[1.07203482 1.15064058] False
[1.04919107 1.09691743] False
[1.04919109 1.09691743] False
[1.04919107 1.09691744] False
[1.05190812 1.10466899] False
[1.05190814 1.10466899] False
[1.05190812 1.104669  ] False
[1.04525088 1.09196528] False
[1.0452509  1.09196528] False
[1.04525088 1.0919653 ] False
[1.02121237 1.0429307 ] False
[1.02121239 1.0429307 ] False
[1.02121237 1.04293072] False
[1.00126109 1.00177624] False
[1.00126111 1.00177624] False
[1.00126109 1.00177626] False
[1.0022548  1.00432578] False
[1.00225481 1.00432578] False
[1.0022548  1.00432579] False
[1.0004773  1.00094749] False
[1.00047732 1.00094749] False
[1.0004773 1.0009475] False
0 35
```

For reference scipy 1.4.1 gives `nfev = 45` and `dup = 15` for the same example.
